### PR TITLE
[FIX] web_editor: change tag between br

### PR DESCRIPTION
**Pull request by deso-odoo**
Current behavior before PR:

The `<br>` tags copied from external editing software's were remained unchanged.

Desired behavior after PR is merged:

Implement breaking element at `<br>` during HTML paste to adress unwanted `<br>`
elements introduced by external editing software's or the OS.

This ensures consistent and correct behaviour for all commands.

Note that this change does not break `<li>`, `<blockquote>`, and `<pre>`
elements.

task-2936891

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 1]
```diff
@@ -3098,6 +3099,20 @@ export class OdooEditor extends EventTarget {
                 fragment.appendChild(p);
             }
         }
+
+        // Split elements containing <br> into seperate elements for each line.
+        const brs = fragment.querySelectorAll('br');
+        for (let br of brs) {
+            const nearestBlock = closestBlock(br);
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3106]: change ` const nearestBlock = closestBlock(br);`  to ` const block = closestBlock(br);`

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 2]
```diff
@@ -3098,6 +3099,20 @@ export class OdooEditor extends EventTarget {
...
+            if (!isEmptyBlock(nearestBlock) &&
+                ['P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'PRE'].includes(nearestBlock.nodeName) &&
+                nearestBlock.parentNode.nodeName !== 'LI') {
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3109]: !closestBlock(block, 'LI') will do.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 3]
```diff
@@ -3098,6 +3099,20 @@ export class OdooEditor extends EventTarget {
...
             }
         }
+
+        // Split elements containing <br> into seperate elements for each line.
+        const brs = fragment.querySelectorAll('br');
+        for (let br of brs) {
+            const block = closestBlock(br);
+            if (['P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'PRE'].includes(block.nodeName) &&
+                !block.closest('li')) {
+                let isEmptyLineBr = (block.firstChild.nodeName === 'BR');
```
Comment (shsa-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3109]: `const isEmptyLineBr = block.firstChild.nodeName === 'BR';` seems better

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 4]
```diff
@@ -3159,6 +3160,20 @@ export class OdooEditor extends EventTarget {
                 fragment.appendChild(p);
             }
         }
+
+        // Split elements containing <br> into seperate elements for each line.
+        const brs = fragment.querySelectorAll('br');
+        for (let br of brs) {
+            const block = closestBlock(br);
+            if (['P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'PRE'].includes(block.nodeName) &&
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3168]: I think you could use `paragraphRelatedElements` here. It also includes `BLOCKQUOTE` but I think that's precisely a good thing.

**Review by Zynton — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 4 above)
Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3168]: That was the very reason of not using `paragraphRelatedElements`, no one would like if `BLOCKQUOTE` would break into multiple `BLOCKQUOTE`s

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3168]: The same could be said of `pre` though. I don't really see what makes blockquote so different.

@pierre-pde Can we have your input please?

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3168]: @deso-odoo After talking with @pierre-pde, the decision is to neither include `BLOCKQUOTE` _or_ `PRE` here.

```suggestion
            if (['P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'].includes(block.nodeName) &&
```

**Comment by deso-odoo** [comment]
Hello, @Zynton I have implemented the requested changes. And also I have implemented an approach to remove the superfluous br issue you found.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 5]
```diff
@@ -3173,6 +3174,20 @@ export class OdooEditor extends EventTarget {
                 fragment.appendChild(p);
             }
         }
+
+        // Split elements containing <br> into seperate elements for each line.
+        const brs = fragment.querySelectorAll('br');
+        for (let br of brs) {
+            const block = closestBlock(br);
+            if (['P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'].includes(block.nodeName) &&
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3182]: Please mention as a side-note in your commit message that behavior change for 'PRE'.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js [hunk 6]
```diff
@@ -293,6 +294,13 @@ export const editorCommands = {
         }
 
         currentNode = lastChildNode || currentNode;
+        if (
+            currentNode.nextSibling &&
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js:298]: Is it impossible for `currentNode` to itself be a br? Because `<p>ab<br><br></p>` is legit.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js [hunk 7]
```diff
@@ -768,6 +761,150 @@ describe('Paste', () => {
...
     });
+    describe('Simple html elements containing <br>', () => {
+        describe('breaking <br> elements', () => {
+            it('should split h1 with <br> into seperate h1 elements', async () => {
+                await testEditor(BasicEditor, {
+                    contentBefore: '<p>[]<br></p>',
+                    stepFunction: async editor => {
+                        await pasteHtml(editor, '<h1>abc<br>def<br>ghi<br>jkl</h1>');
+                    },
+                    contentAfter: '<p>abc</p><h1>def</h1><h1>ghi</h1><p>jkl[]<br></p>',
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js]: This is an incorrect result which gets fixed with your next commit. It would be better to swap your two commits to you can directly write the correct test case.

**Review by Zynton — CHANGES_REQUESTED** [review]
Good job, we're almost there :+1:

File: addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js (hunk 6 above)
Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js:298]: `currentNode` can be a `br`. Could you please elaborate the test case you are referencing to?
What should be the contentAfter here:
1. ContentBefore: `<p>[]<br></p>`,
    pasteHtml: `<p>ab<br></p>`,
    ContentAfter: ?
    
2. ContentBefore: `<p>[]<br></p>`,
    pasteHtml: `<p>ab<br><br></p>`,
    ContentAfter: ?

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js:298]: 1. ContentBefore: `<p>[]<br></p>`,
    pasteHtml: `<p>ab<br></p>`,
    ContentAfter: `<p>ab</p>`

2. ContentBefore: `<p>[]<br></p>`,
    pasteHtml: `<p>ab<br><br></p>`,
   ContentAfter: `<p>ab<br><br></p>`

The br is `<p>ab<br></p>` serves no function. The two brs in `<p>ab<br><br></p>` create a new line after "ab". You can try this yourself to see what it looks like.

Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/commands/commands.js:306]: Hello, @Zynton After the fix in the second commit, copying and pasting `<p>ab<br></p>` from outside Odoo was working as expected it removed the last redundant `<br>`. But however copying `<p>ab<br></p>` from editor itself, the `<br>` was not removed on `insert`. I have implemented an approach to remove that `<br>` in the third commit. And about `<p>ab<br><br></p>` it works fine there is no issue with it.

PS: This `<br>` with `<p>ab<br></p>` is not related to the one we fixed in first commit.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js [hunk 8]
```diff
@@ -79,6 +79,33 @@ describe('insert HTML', () => {
...
+                contentAfter: '<p>abc[]</p>',
+            });
+        });
+        it('should not remove the trailing single <br> when pasting between text', async () => {
+            await testEditor(BasicEditor, {
+                contentBefore: '<p>ab[]cd</p>',
+                stepFunction: async editor => {
+                    await editor.execCommand('insert', parseHTML('<p>efg<br></p>'));
+                },
+                contentAfter: '<p>abefg<br>[]cd</p>',
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:97]: That's a strange behavior actually because `<p>efg<br></p>` doesn't have a visible line break (hence your previous test) so I'd expect `<p>abefg[]cd</p>`.

@sebgeelen what do you think?

Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:97]: It's probably correct since it's the kind of html you will get in the clipboard when you copy this : 
```
abcd[efg
]
hij
```

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:97]: Really? But this test and the one above it can't both be true at the same time, can they? And the one above seems right to me.
For what it's worth, I just tested in this `textarea` and it's inconsistent:
Pasting 
image
from your comment in here never includes the line break, while 
image
does.

Comment (sebgeelen) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:97]: Ok, 

The silver lining here is that this :  ``<p>efg<br></p>`` could append (in the paste clipboard).
So IMO, the test result make sense.

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:97]: OK so then the previous test doesn't. My bad because it's me who suggested it. Sorry @deso-odoo :pray: 

I tested in a bare contenteditable and indeed copying from `<p>[abc<br>]def</p>` gives the html `<p>abc<br></p>` in the clipboard, which makes contenteditable behave inconsistently since pasting it at `<p>hel[]lo</p>` will give  `<p>helabc<br>[]lo</p>` (which shows a visible line break) but pasting at `<p>hello[]</p>` will give `<p>helloabc<br>[]</p>` (which doesn't show a visible break so that the cursor looks like it's attached to the `c`).

Conclusion: even though the `br` at the end of `<p>abc<br></p>` isn't visible, in the clipboard it signifies a visible line break, therefore pasting `<p>abc<br></p>` should always show a visible line break at the end.

In other words:
```js
before: '<p>[]<br></p>'
paste('<p>abc<br></p>') // pasted html ends with 1 visible line break
after: '<p>abc<br>[]<br></p>' // 1 visible line break at the end (selection can be after the last, it makes no difference)
```
```js
before: '<p>ab[]cd</p>'
paste('<p>efg<br></p>') // pasted html ends with 1 visible line break
after: '<p>abefg<br>[]cd</p>' // 1 visible line break in the middle
```
```js
before: '<p>[]<br></p>'
paste('<p>abc<br><br></p>')  // pasted html ends with 2 visible line breaks
after: '<p>abc<br><br>[]<br></p>' // 2 visible line breaks at the end (selection can be after the last, it makes no difference)
```

Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:95]: @Zynton Then we might not need the third commit, it can be done with the first commit itself. I have made the changes. It should be good now. :slightly_smiling_face:

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js [hunk 9]
```diff
@@ -768,6 +761,159 @@ describe('Paste', () => {
...
+                    contentAfter: '<p>abc</p><p>def[]</p>',
+                });
+            });
+            it('should split <br> elements with empty lines', async () => {
+                await testEditor(BasicEditor, {
+                    contentBefore: '<p>[]<br></p>',
+                    stepFunction: async editor => {
+                        await pasteHtml(editor, '<p><br><br><br></p>');
+                    },
+                    contentAfter: '<p><br></p><p><br></p><p><br>[]<br></p>',
```
Comment (deso-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js:874]: @Zynton You might not like this two br's at end here, but it will be fixed to contentAfter: `(<p><br></p><p><br></p><p>[]<br></p>)` with https://github.com/odoo/odoo/pull/145220

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js:874]: Didn't we establish [here](https://github.com/odoo/odoo/pull/136743#discussion_r1492538848) that 3 line breaks in the clipboard means we should insert 3 visible line breaks in the end? We established that for `<p>a<br><br><br></p>` but not specifically for `<p><br><br><br></p>` so I'm not 100% sure. But if that's the case, this is wrong since you're saying the other PR will remove the extra `<br>` and leave us short of one line break.

Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js:874]: So I tested it again on bare contenteditable [here](https://jsfiddle.net/f3gx5vzp/), and we get `<p><br><br><br></p>` when we copy the content from the start of fourth line meaning
`<p>`
`  [<br>`
`  <br>`
`  <br>`
`  ]<br>`
`</p>`(I copied 4 lines). So right now i was breaking `<p><br><br><br></p>` to `<p><br></p><p><br></p><p><br></p>` rather I should break it to `<p><br></p><p><br></p><p><br></p><p><br></p>`. So the contentAfter would be `<p><br></p><p><br></p><p><br></p><p>[]<br></p>`. Also I'll bring the [commit](https://github.com/odoo/odoo/pull/145220/commits/4df5277c1d17963f1639a500edabb96e0d82be7c) I referenced from another [PR](https://github.com/odoo/odoo/pull/145220) here so its all at one place.

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js:874]: Yes. You have 1 paragraph and you're inserting 3 new paragraphs so you should have 4 at the end.

@dmo-odoo @sebgeelen if you want to confirm...

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js:874]: In case 2 we want to add 3 line breaks right? But here we only add 2 since `<p>a<br><br><br></p>` only shows 2 visible lines after the letter "a":
image
But the version with paragraph breaks is correct.

In case 3 I'm afraid the same is true.

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/copyPaste.test.js:874]: I'm not sure I understand what you mean so we'll need to discuss it together on monday.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 10]
```diff
@@ -3234,6 +3235,22 @@ export class OdooEditor extends EventTarget {
                 fragment.appendChild(p);
             }
         }
+
+        // Split elements containing <br> into seperate elements for each line.
+        const brs = fragment.querySelectorAll('br');
+        for (const br of brs) {
+            const block = closestBlock(br);
+            if (
+                ['P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'].includes(block.nodeName) &&
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3244]: Shouldn't this be `paragraphRelatedElements` ?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 11]
```diff
@@ -3234,6 +3235,22 @@ export class OdooEditor extends EventTarget {
...
+                !block.closest('li')
+            ) {
+                const isEmptyLineBreak = block.firstChild.nodeName === 'BR';
+                const splitElement = splitAroundUntil(br, block);
+                if (isEmptyBlock(splitElement) && !isEmptyLineBreak) {
+                    splitElement.remove();
+                }
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3251]: I don't understand this condition nor the case it wants to cover.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 10 above)
Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3244]: :smile:  See https://github.com/odoo/odoo/pull/136743#discussion_r1469571048

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 11 above)
Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3251]: Let's assume I have `<p>abc<br><br>def</p>`. The first `<br>` serves as a line break, while the second `<br>` indicates an empty line.

If I split around the `<p>` tag at the first `<br>`, it would become `<p>abc</p><p><br></p><p><br>def<p>`. I need to remove the `<p>` element containing the first `<br>`. However, for the second `<br>` I don't want to remove the split element, as it signifies an empty line. And how can I determine if a `<br>` represents an empty line, One way is to check if its closest block's `firstChild` is `<br>`.

See this test case: https://github.com/odoo/odoo/pull/136743/files#diff-ed334b1f9f243d6fb9d00527d62af2dd515fd4ea0442c9c89d811c0caa1ad19bR842-R848

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3251]: I see, thanks. I suggest to add a comment on top of this code then, explaining this, as it's impossible to figure out without a deep knowledge of what is going on... ^^'

Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3256]: Added the comment. Also, I realized if I `splitAroundUntil` br till its `closestBlock` the splitElement will always be an emptyBlock :eyes: so I actually did not need to check that, Therefore I changed the if condition from `if (isEmptyBlock(splitElement) && !isEmptyLineBreak)` to `if (!isEmptyLineBreak)`

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3256]: Thanks! Now that I have a better understanding of the issue, I suggest this:
```js
                // A linebreak at the beginning of a block is an empty line.
                const isEmptyLine = block.firstChild.nodeName === 'BR';
                // Split blocks around it until only the BR remains in the block.
                const remainingBrContainer = splitAroundUntil(br, block);
                // Remove the container unless it represented an empty line.
                if (!isEmptyLine) {
                    remainingBrContainer.remove();
                }
```

**Comment by dmo-odoo** [comment]
Great work, thanks! 

@robodoo r+
